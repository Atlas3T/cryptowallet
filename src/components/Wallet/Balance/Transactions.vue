<template>
  <section class="recent-transactions">
    <q-tabs color="secondary">
      <q-tab
        slot="title"
        :label="$t('all')"
        default
        name="tab-1"
      />
      <q-tab
        slot="title"
        :label="$t('sent')"
        name="tab-2"
      />
      <q-tab
        slot="title"
        :label="$t('received')"
        name="tab-3"
      />
      <q-tab-pane name="tab-1">
        <TransactionsList/>
      </q-tab-pane>
      <q-tab-pane name="tab-2">
        <TransactionsList filter="sent"/>
      </q-tab-pane>
      <q-tab-pane name="tab-3">
        <TransactionsList filter="received"/>
      </q-tab-pane>
    </q-tabs>
  </section>
</template>

<script>
import TransactionsList from './TransactionsList.vue';

export default {
  name: 'Transactions',
  components: {
    TransactionsList,
  },
};
</script>

<style>
  .recent-payments-scroll-area {
    width: auto;
    height: 100%;
  }

  .recent-transactions {
    height: 55vh;
    border-top: 2px solid #09233a;
    position: relative;
    background: #16324a;
  }

  .scroll-content {
    overflow: hidden;
    padding: 0 1rem;
  }

  .q-tabs {
    height: 100%;
    display: flex;
  }

  .q-tabs-panes {
    flex: 1;
    display: flex;
  }

  .q-tab-pane {
    flex: 1;
    font-size: 0.75rem;
  }
</style>
