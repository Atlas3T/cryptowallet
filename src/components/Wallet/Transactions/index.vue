<template>
  <section class="recent-transactions">
    <q-tabs
      v-model="tab"
      color="blueish"
      inverted
    >
      <q-tab
        :label="$t('all')"
        default
        name="all"
      />
      <q-tab
        :label="$t('sent')"
        name="sent"
      />
      <q-tab
        :label="$t('received')"
        name="received"
      />
    </q-tabs>

    <q-tab-panels
      v-model="tab"
      animated
    >
      <q-tab-panel
        dark
        name="all"
      >
        <TransactionsList :wallet="wallet" />
      </q-tab-panel>

      <q-tab-panel name="sent">
        <TransactionsList
          :wallet="wallet"
          filter="sent"
        />
      </q-tab-panel>

      <q-tab-panel name="received">
        <TransactionsList
          :wallet="wallet"
          filter="received"
        />
      </q-tab-panel>
    </q-tab-panels>
  </section>
</template>

<script>
import TransactionsList from '@/components/Wallet/TransactionsList';

export default {
  name: 'Transactions',
  components: {
    TransactionsList,
  },
  props: {
    wallet: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      tab: 'all',
    };
  },
};
</script>

<style>
.q-tab-panels {
  background: #0a2d3e
}

.q-item__label--caption {
  color: lightgrey;
 }

.recent-payments-scroll-area {
  width: auto;
  height: 100%;
}

.recent-transactions {
  height: 100%;
  position: relative;
}

.recent-transactions .q-tab-panels {
  height: calc(100% - 48px);
  position: relative;
}

.scroll-content {
  overflow: hidden;
  padding: 0 1rem;
}

.q-tabs, .q-tabs__content {
  width: 100%!important;
}

.q-tab-label {
  font-family: Montserrat-SemiBold;
}

.q-tabs-head {
  background: whitesmoke;
}

.q-timeline {
  margin: 0;
}
</style>
