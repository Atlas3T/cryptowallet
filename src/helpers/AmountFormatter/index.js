/**
 * @TODO James Konrad
 * I've mocked some exchange rates below for the amount() function to use
 * Would make sense to put them somewhere in the Vuex store
 */

import numeral from 'numeral';

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GBP',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GBP',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GBP',
      rate: 9999,
    },
  ],
};

export default class AmountFormatter {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  getFormatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${this.currency.symbol}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `- ${formatted}`;
      if (amount > 0) return `+ ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency.code);
    if (rate) return this.amount * rate;
    return this.amount;
  }
}
