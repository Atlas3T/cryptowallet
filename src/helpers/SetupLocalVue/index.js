import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
