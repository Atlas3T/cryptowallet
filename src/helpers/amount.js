/**
 * @TODO James Konrad
 * mocked data below
 */

import numeral from 'numeral';

const currencySymbols = {
  USD: '$',
  GPB: 'Â£',
};

const exchangeRates = {
  BTC: [
    {
      currency: 'USD',
      rate: 6431.14,
    },
    {
      currency: 'GPB',
      rate: 4977.06,
    },
  ],
  ETH: [
    {
      currency: 'USD',
      rate: 201.76,
    },
    {
      currency: 'GPB',
      rate: 156.14,
    },
  ],
  ATL: [
    {
      currency: 'USD',
      rate: 22222,
    },
    {
      currency: 'GPB',
      rate: 9999,
    },
  ],
};

export default class Amount {
  constructor(options) {
    this.amount = options.amount;
    this.coin = options.coin;
    this.currency = options.currency;
    this.format = options.format;
    this.withCurrencySymbol = options.withCurrencySymbol;
    this.prependPlusOrMinus = options.prependPlusOrMinus;
  }

  /**
   * Formats a Number and returns a String
   * @return {String}
   */
  get formatted() {
    let { amount } = this;
    if (this.coin && this.currency) {
      amount = this.coinToCurrency(this.amount, this.coin, this.currency);
    }

    let formatted = `${numeral(Math.abs(amount)).format(this.format)}`;
    if (this.currency) formatted = `${currencySymbols[this.currency]}${formatted}`;

    if (this.prependPlusOrMinus) {
      if (amount < 0) return `&#45; ${formatted}`;
      if (amount > 0) return `&#43; ${formatted}`;
    }

    return formatted;
  }

  /**
   * Converts a coin into a currency
   * @return {Number}
   */
  coinToCurrency() {
    const { rate } = exchangeRates[this.coin].find(item => item.currency === this.currency);
    return this.amount * rate;
  }
}
