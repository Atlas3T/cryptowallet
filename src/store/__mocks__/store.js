import Vue from 'vue';
import Vuex from 'vuex';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from '@atlascity/vuex-orm-plugin-lokijs';

import wallet from '@/store/wallet';
import contacts from '@/store/contacts/__mocks__/contacts.js';
import search from '@/store/search/__mocks__/search.js';
import settings from '@/store/settings/__mocks__/settings.js';
import setup from '@/store/setup/__mocks__/setup.js';
import qrcode from '@/store/qrcode/__mocks__/qrcode.js';
import modals from '@/store/modals/__mocks__/modals.js';

// entities
import Account from '@/store/wallet/entities/account';
import Address from '@/store/wallet/entities/address';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import KeyPair from '@/store/wallet/entities/keyPair';
import Prices from '@/store/prices';
import LatestPrice from '@/store/latestPrice';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
  Wallet,
];

const options = {
  env: 'browser',
};

VuexORM.use(VuexORMLoki, { database, options });

export const actions = {};
export const getters = {};
export const mutations = {};
export const state = {};

export function createMocks(custom = {
  getters: {}, mutations: {}, actions: {}, state: {},
}) {
  /**
   * mock store modules.
   */
  const modules = {
    wallet,
    search,
    settings,
    contacts,
    setup,
    qrcode,
    modals,
  };

  let mockGettersCombined = {};
  let mockMutationsCombined = {};
  let mockActionsCombined = {};
  let mockStateCombined = {};

  Object.entries(modules).forEach(([name, module]) => {
    mockGettersCombined = Object.assign({}, mockGettersCombined, module.getters);
    mockMutationsCombined = Object.assign({}, mockMutationsCombined, module.mutations);
    mockActionsCombined = Object.assign({}, mockActionsCombined, module.actions);
    mockStateCombined = Object.assign({}, mockStateCombined, module.state);

    if (custom.state && custom.state[name]) {
      module.state = Object.assign(module.state, custom.state[name]);
    }

    modules[name] = {
      namespaced: true,
      getters: Object.assign({}, getters, module.getters),
      mutations: Object.assign({}, mutations, module.mutations),
      actions: Object.assign({}, actions, module.actions),
      state: Object.assign({}, state, module.state),
    };
  });

  mockGettersCombined = Object.assign({}, mockGettersCombined, custom.getters);
  mockMutationsCombined = Object.assign({}, mockMutationsCombined, custom.mutations);
  mockActionsCombined = Object.assign({}, mockActionsCombined, custom.actions);
  mockStateCombined = Object.assign({}, mockStateCombined, custom.state);
  return {
    getters: mockGettersCombined,
    mutations: mockMutationsCombined,
    actions: mockActionsCombined,
    state: mockStateCombined,
    store: new Vuex.Store({
      getters: mockGettersCombined,
      mutations: mockMutationsCombined,
      actions: mockActionsCombined,
      state: mockStateCombined,
      modules,
      plugins: [VuexORM.install(database)],
    }),
  };
}

export const { store } = createMocks();
