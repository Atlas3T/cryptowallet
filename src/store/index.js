/*eslint-disable*/

import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';
import VuexORM from '@vuex-orm/core';
import VuexORMLoki from 'vuex-orm-lokijs';

// import entities.
import Account from './wallet/entities/account';
import Address from './wallet/entities/address';
import Tx from './wallet/entities/tx';
import Utxo from './wallet/entities/utxo';
import Wallet from './wallet/entities/wallet';
import Coin from './wallet/entities/coin';
import KeyPair from './wallet/entities/keyPair';
import Prices from './prices';
import LatestPrice from './latestPrice';

// import modules.
import payments from './payments';
import search from './search';
import settings from './settings';
import setup from './setup';
import contacts from './contacts';

Vue.use(Vuex);

// Setup ORM database.
const database = new VuexORM.Database();

database.register(Account, {});
database.register(Address, {});
database.register(Tx, {});
database.register(Utxo, {});
database.register(Wallet, {});
database.register(Coin, {});
database.register(KeyPair, {});
database.register(Prices, {});
database.register(LatestPrice, {});

Vue.prototype.encryptedModels = [
  Account,
];

window.Tx = Tx;
window.Utxo = Utxo;
window.Address = Address;
window.Wallet = Wallet;
window.LatestPrice = LatestPrice;
window.Prices = Prices;
window.Account = Account;
window.Coin = Coin;

// Setup persistant storage.
const vuexLocal = new VuexPersistence({
  supportCircular: true,
  storage: window.localStorage,
  modules: [settings],
});


const options = {
  env: 'browser',
};

function hydrationCompletedCallback() {
  setTimeout(() => {
    store.dispatch('settings/setLoading', false);
  }, 1000);
}

VuexORM.use(VuexORMLoki, { database, options, hydrationCompletedCallback });


/**
 * Create CryptoWallet Vuex store obj.
 * @type {Store<any>}
 */
const store = new Vuex.Store({
  modules: {
    payments,
    search,
    settings,
    setup,
    contacts,
  },
  plugins: [VuexORM.install(database)],
});

Vue.prototype.activeWallets = {};

// @todo remove line below (used for development purposes only)
window.store = store;

export default store;
