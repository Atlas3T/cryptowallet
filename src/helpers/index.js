import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
