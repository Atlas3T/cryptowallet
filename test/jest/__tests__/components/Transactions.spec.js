import { shallowMount } from '@vue/test-utils';
import Transactions from '@/components/Wallet/Transactions';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';

describe('Transactions.vue', () => {
  let storeMocks;
  let wrapper;
  // let store;
  let router;
  const propsData = {
    wallet: {
      account_id: 1,
      displayName: 'Bitcoin',
      sdk: 'Bitcoin',

    },
  };

  function wrapperInit(options) {
    return shallowMount(Transactions, options);
  }

  function storeInit(custom) {
    storeMocks = createStoreMocks(custom);
    router = createRouter(storeMocks.store);
    router.push({ path: '/' });
    wrapper = wrapperInit({
      i18n, router, localVue, store: storeMocks.store, propsData,
    });
    // store = wrapper.vm.$store;
  }

  beforeEach(() => { return storeInit(); });

  it('renders and matches snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders a secion with .recent-transactions class', () => {
    expect(wrapper.contains('section.recent-transactions')).toBe(true);
  });

  it('renders tabs: All, Sent, Received correctly', () => {
    expect(wrapper.findAll('q-tab-stub').length).toBe();
    expect(wrapper.contains('q-tab-stub[default="true"][name="tab-1"]')).toBe(true);
    expect(wrapper.contains('q-tab-stub[label="Sent"][name="tab-2"]')).toBe(true);
    expect(wrapper.contains('q-tab-stub[label="Received"][name="tab-3"]')).toBe(true);

    expect(wrapper.contains('q-tabpane-stub[name="tab-1"] transactionslist-stub')).toBe(true);
    expect(wrapper.contains('q-tabpane-stub[name="tab-2"] transactionslist-stub[filter="sent"]')).toBe(true);
    expect(wrapper.contains('q-tabpane-stub[name="tab-3"] transactionslist-stub[filter="received"]')).toBe(true);
  });
});
