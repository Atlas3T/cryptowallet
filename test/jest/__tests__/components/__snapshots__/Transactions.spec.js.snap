// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions.vue renders and matches snapshot 1`] = `
<section
  class="recent-transactions"
>
  <q-tabs-stub
    align="center"
    breakpoint="600"
    color="blueish"
    inverted=""
    value="all"
  >
    <q-tab-stub
      default=""
      label="All"
      name="all"
      ripple="true"
    />
     
    <q-tab-stub
      label="Sent"
      name="sent"
      ripple="true"
    />
     
    <q-tab-stub
      label="Received"
      name="received"
      ripple="true"
    />
  </q-tabs-stub>
   
  <q-tab-panels-stub
    animated="true"
    transitionnext="slide-left"
    transitionprev="slide-right"
    value="all"
  >
    <q-tab-panel-stub
      name="all"
    >
      <transactionslist-stub
        filter=""
        wallet="[object Object]"
      />
    </q-tab-panel-stub>
     
    <q-tab-panel-stub
      name="sent"
    >
      <transactionslist-stub
        filter="sent"
        wallet="[object Object]"
      />
    </q-tab-panel-stub>
     
    <q-tab-panel-stub
      name="received"
    >
      <transactionslist-stub
        filter="received"
        wallet="[object Object]"
      />
    </q-tab-panel-stub>
  </q-tab-panels-stub>
</section>
`;
